### PR TITLE
Basic NGSI-LD active measures support

### PR DESCRIPTION
**Pull request by jason-fox**
This PR offers NGSI-LD support for **active measures only** under the following configuration.

```javascript
contextBroker: {
            url: 'https://192.168.1.1:1026',
            ngsiVersion: 'ld',
            jsonLdContext: 'http://context.json-ld'
        },
```

The existing NGSI-v2 support remains as default under `ngsiVersion: 'v2' `

The NGSI-LD support matches the decision trees  for NGSI v1/v2 - where necessary a third option has been added and the necessary functionality **duplicated**.

The set of tests from NGSI v2 have been duplicated and the expectations amended where necessary.

Commands, Subscriptions and Lazy Attribute Support have not been added - disabled test stubs have been created to allow them to be tested once the code base has been updated.

### CODE REVIEW

**Comment by jason-fox** [comment]
The tests cases are *copies* of the NGSI-v2 tests. The disabled tests are for commands and lazy attributes and therefore *will not work* until the remaining work is done. 

Given that the tests are all new,  I'm wondering - *for the tests only* should I:

a) Make minimal duplicate and amend changes (as is)
b) Prettify the changed test files - to reduce the scope of #753 
c) Update the tests to run using ES6 syntax - to reduce the scope of #831

**Comment by jason-fox** [comment]
The command stubs account for the 2% drop in code coverage.

File: CHANGES_NEXT_RELEASE [hunk 1]
```diff
@@ -1,4 +1,5 @@
 Refresh Documentation
 Add NGSIv2 metadata support to device provisioned attributes
 Fix: Error message when sending measures with unknown/undefined attribute
-Add Null check within executeWithSecurity() to avoid crash (#829)
\ No newline at end of file
+Add Null check within executeWithSecurity() to avoid crash (#829)
+Add NGSIv2 metadata support to attributeAlias plugin.
```
Comment (fgalan) [inline CHANGES_NEXT_RELEASE]: This line seems to belong to PR #839. Thus, I understand that jason-fox:feature/ngsi-ld-measure is a branch of jason-fox:feature/add-v2-metadata.

PR #839 should be merged before, then this PR updated.

**Comment by fgalan** [comment]
> Given that the tests are all new, I'm wondering - _for the tests only_ should I:
> 
> a) Make minimal duplicate and amend changes (as is)
> b) Prettify the changed test files - to reduce the scope of #753
> c) Update the tests to run using ES6 syntax - to reduce the scope of #831

Given that these .js files are entirely new, maybe it makes sense to do so.

**Comment by fgalan** [comment]
> The command stubs account for the 2% drop in code coverage.

Not a big issue from my point of view. The drop will be eventually recovered.

File: CHANGES_NEXT_RELEASE (hunk 1 above)
Comment (jason-fox) [reply CHANGES_NEXT_RELEASE]: Correct - this PR is dependent upon  and extends the metadata change.

**Comment by jason-fox** [comment]
> Given that these .js files are entirely new, maybe it makes sense to do so.

Fixed  (new tests only) 40875e3 - I also corrected the copyrights for the new files to 2020

File: CHANGES_NEXT_RELEASE (hunk 1 above)
Comment (fgalan) [reply CHANGES_NEXT_RELEASE]: PR #839 has been merged. This PR branch should be upgrade, then this diff will disappear.

Comment (jason-fox) [reply CHANGES_NEXT_RELEASE]: I've merge master which should reduce the diff - Fixed with dd81751

File: lib/services/devices/deviceService.js [hunk 2]
```diff
@@ -143,6 +143,41 @@ function createInitialEntityHandlerNgsi2(deviceData, newDevice, callback) {
     };
 }
 
+/**
+ * Creates the response handler for the initial entity creation request using NGSIv2.
```
Comment (fgalan) [inline lib/services/devices/deviceService.js]:
```suggestion
 * Creates the response handler for the initial entity creation request using NGSI-LD.
```

File: lib/commonConfig.js [hunk 3]
```diff
@@ -91,7 +91,8 @@ function processEnvironmentVariables() {
              'IOTA_APPEND_MODE',
              'IOTA_POLLING_EXPIRATION',
              'IOTA_POLLING_DAEMON_FREQ',
-             'IOTA_MULTI_CORE'
+             'IOTA_MULTI_CORE',
+             'IOTA_JSON_LD_CONTEXT'
```
Comment (fgalan) [inline lib/commonConfig.js]: New env var should be described in documentation.

File: lib/commonConfig.js [hunk 4]
```diff
@@ -91,7 +91,8 @@ function processEnvironmentVariables() {
              'IOTA_APPEND_MODE',
```
Comment (fgalan) [inline lib/commonConfig.js]: CHANGES_NEXT_RELEASE entry should be provided for this PR.

File: lib/services/devices/deviceService.js [hunk 5]
```diff
@@ -310,6 +348,66 @@ function executeWithSecurity(requestOptions, deviceData, callback) {
         });
 }
 
+
+/**
+ * Creates the initial entity representing the device in the Context Broker using NGSIv2.
```
Comment (fgalan) [inline lib/services/devices/deviceService.js]:
```suggestion
 * Creates the initial entity representing the device in the Context Broker using NGSI-LD.
```

File: lib/services/devices/deviceService.js [hunk 6]
```diff
@@ -522,6 +621,64 @@ function updateEntityNgsi2(deviceData, updatedDevice, callback) {
     }
 }
 
+
+
+/**
+ * Updates the entity representing the device in the Context Broker using NGSIv2.
```
Comment (fgalan) [inline lib/services/devices/deviceService.js]:
```suggestion
 * Updates the entity representing the device in the Context Broker using NGSIv2.
```

Three times :) Please, do an overall check for this.

File: lib/services/ngsi/ngsiService.js [hunk 7]
```diff
@@ -397,6 +485,83 @@ function getMetaData(typeInformation, name, metadata){
...
+    return !isNaN(value) && value.toString().indexOf('.') !== -1;
+}
+
+function orBlank(value){
+    return isNaN(value) ?  {'@type':  'Intangible', '@value':  null} : value;
+}
+
+
+
+function convertNGSIv2ToLD(attr){
```
Comment (fgalan) [inline lib/services/ngsi/ngsiService.js]: JavaDoc should be provided?

File: lib/services/ngsi/ngsiService.js [hunk 8]
```diff
@@ -397,6 +485,83 @@ function getMetaData(typeInformation, name, metadata){
...
+    if (attr.metadata){
+        Object.keys(attr.metadata).forEach(function(key) {
+          obj[key] = convertNGSIv2ToLD(attr.metadata[key]);
+        });
+    }
+
+    return obj;
+}
+
+function formatAsNGSILD(json){
```
Comment (fgalan) [inline lib/services/ngsi/ngsiService.js]: JavaDoc should be provided?

File: lib/services/ngsi/ngsiService.js [hunk 9]
```diff
@@ -584,14 +741,172 @@ function sendUpdateValueNgsi2(entityName, attributes, typeInformation, token, ca
...
+}
+
+
+
+
+
+
+
+
+
```
Comment (fgalan) [inline lib/services/ngsi/ngsiService.js]: Big bunch of empty blank lines... intentional?

File: test/.jshintrc [hunk 10]
```diff
@@ -30,6 +30,6 @@
     },
     "predef":
         [
-            "describe", "beforeEach", "afterEach", "it"
+            "describe", "beforeEach", "afterEach", "it", "xdescribe", "xit"
```
Comment (fgalan) [inline test/.jshintrc]: I have check there are 26 xdescribe() statement and 4 xit() statement added in this PR.

I understand that the work in sucesive PRs within the NGSI-LD line will be re-enable them, isn't it?

Comment (jason-fox) [reply test/.jshintrc]: Yes

File: lib/services/ngsi/ngsiService.js (hunk 9 above)
Comment (jason-fox) [reply lib/services/ngsi/ngsiService.js]: Fixed b41a18a  (if only this tidying could be done automatically)

File: lib/services/ngsi/ngsiService.js (hunk 8 above)
Comment (jason-fox) [reply lib/services/ngsi/ngsiService.js]: Fixed b41a18a

File: lib/services/ngsi/ngsiService.js (hunk 7 above)
Comment (jason-fox) [reply lib/services/ngsi/ngsiService.js]: Fixed b41a18a

File: lib/services/devices/deviceService.js (hunk 6 above)
Comment (jason-fox) [reply lib/services/devices/deviceService.js]: Fixed b41a18a

File: lib/services/devices/deviceService.js (hunk 5 above)
Comment (jason-fox) [reply lib/services/devices/deviceService.js]: Fixed b41a18a

File: lib/commonConfig.js (hunk 4 above)
Comment (jason-fox) [reply lib/commonConfig.js]: Fixed b41a18a

File: lib/commonConfig.js (hunk 3 above)
Comment (jason-fox) [reply lib/commonConfig.js]: Fixed b41a18a  - Minimal Documentation for now.

File: lib/services/devices/deviceService.js (hunk 2 above)
Comment (jason-fox) [reply lib/services/devices/deviceService.js]: Fixed b41a18a

File: test/.jshintrc (hunk 10 above)
Comment (fgalan) [reply test/.jshintrc]: Ok, NTC

File: lib/services/devices/registrationUtils.js [hunk 11]
```diff
@@ -329,31 +378,145 @@ function sendRegistrationsNgsi2(unregister, deviceData, callback) {
...
+        deviceData,
+        createRegistrationHandlerNgsi2(unregister, deviceData, callback));
+
+    
+    
+}
+
+
+/**
+ * Sends a Context Provider registration or unregistration request to the Context Broker using NGSIv2.
```
Comment (fgalan) [inline lib/services/devices/registrationUtils.js]: Another one ;)

```suggestion
 * Sends a Context Provider registration or unregistration request to the Context Broker using NGSI-LD.
```

File: lib/services/ngsi/ngsiService.js [hunk 12]
```diff
@@ -193,6 +193,88 @@ function generateNGSI2OperationHandler(operationName, entityName, typeInformatio
     };
 }
 
+
+/**
+ * Generate an operation handler for NGSIv2-based operations (query and update). The handler takes care of identifiying
```
Comment (fgalan) [inline lib/services/ngsi/ngsiService.js]:
```suggestion
 * Generate an operation handler for NGSI-LD-based operations (query and update). The handler takes care of identifiying
```

File: lib/services/ngsi/ngsiService.js [hunk 13]
```diff
@@ -584,14 +761,164 @@ function sendUpdateValueNgsi2(entityName, attributes, typeInformation, token, ca
...
             request(options,
                 generateNGSI2OperationHandler('update', entityName, typeInformation, token, options, callback));
         }
     });
 }
 
+
+/**
+ * Makes an update in the Device's entity in the context broker, with the values given in the 'attributes' array. This
+ * array should comply to the NGSIv2's attribute format.
```
Comment (fgalan) [inline lib/services/ngsi/ngsiService.js]:
```suggestion
 * array should comply to the NGSI-LD's attribute format.
```

File: lib/services/ngsi/subscriptionService.js [hunk 14]
```diff
@@ -245,6 +245,58 @@ function subscribeNgsi2(device, triggers, content, callback) {
         device, createSubscriptionHandlerNgsi2(device, triggers, store, callback));
 }
 
+
+/**
+ * Makes a subscription for the given device's entity using NGSIv2, triggered by the given attributes.
```
Comment (fgalan) [inline lib/services/ngsi/subscriptionService.js]:
```suggestion
 * Makes a subscription for the given device's entity using NGSIv2, triggered by the given attributes.
```

Still too many of them... an overall review should be done.

Comment (jason-fox) [reply lib/services/ngsi/subscriptionService.js]: Fixed a027234

File: lib/services/ngsi/ngsiService.js (hunk 13 above)
Comment (jason-fox) [reply lib/services/ngsi/ngsiService.js]: Fixed a027234

File: lib/services/ngsi/ngsiService.js (hunk 12 above)
Comment (jason-fox) [reply lib/services/ngsi/ngsiService.js]: Fixed a027234

File: lib/services/devices/registrationUtils.js (hunk 11 above)
Comment (jason-fox) [reply lib/services/devices/registrationUtils.js]: Fixed a027234

File: lib/services/ngsi/subscriptionService.js (hunk 14 above)
Comment (fgalan) [reply lib/services/ngsi/subscriptionService.js]: This one was not fixed at the end (my fault: my suggestion was not correct in my first comment). Now it should be ok:

```suggestion
 * Makes a subscription for the given device's entity using NGSI-LD, triggered by the given attributes.
```

File: lib/services/northBound/contextServer.js [hunk 15]
```diff
@@ -957,6 +1177,79 @@ function handleNotificationNgsi1(req, res, next) {
     }
 }
 
+function handleNotificationNgsiLD(req, res, next) {
```
Comment (fgalan) [inline lib/services/northBound/contextServer.js]: Missing Javadoc ?

File: lib/services/northBound/contextServer.js [hunk 16]
```diff
@@ -1119,6 +1412,21 @@ function queryErrorHandlingNgsi2(error, req, res, next) {
     });
 }
 
+function queryErrorHandlingNgsiLD(error, req, res, next) {
```
Comment (fgalan) [inline lib/services/northBound/contextServer.js]: Missing JavaDoc?

File: lib/services/northBound/contextServer.js [hunk 17]
```diff
@@ -1159,6 +1467,21 @@ function updateErrorHandlingNgsi2(error, req, res, next) {
     });
 }
 
+function updateErrorHandlingNgsiLD(error, req, res, next) {
```
Comment (fgalan) [inline lib/services/northBound/contextServer.js]: Missing JavaDoc?

Comment (jason-fox) [reply lib/services/northBound/contextServer.js]: Fixed c70d9ed

File: lib/services/northBound/contextServer.js (hunk 16 above)
Comment (jason-fox) [reply lib/services/northBound/contextServer.js]: Fixed c70d9ed

File: lib/services/northBound/contextServer.js (hunk 15 above)
Comment (jason-fox) [reply lib/services/northBound/contextServer.js]: Fixed c70d9ed

File: lib/services/ngsi/subscriptionService.js (hunk 14 above)
Comment (jason-fox) [reply lib/services/ngsi/subscriptionService.js]: Fixed c70d9ed

File: lib/services/northBound/contextServer.js (hunk 16 above)
Comment (fgalan) [reply lib/services/northBound/contextServer.js]: Maybe github diff is fooling me, but it seems JavaDoc was not added here, despite the "Fixed ..." mark.

imagen

Can you check, please?

File: lib/services/northBound/contextServer.js (hunk 17 above)
Comment (fgalan) [reply lib/services/northBound/contextServer.js]: Same here

File: lib/services/northBound/contextServer.js (hunk 15 above)
Comment (jason-fox) [reply lib/services/northBound/contextServer.js]: Fixed and Propagated (contextServer isn't the focus of Measures) : a5c650a

File: lib/services/northBound/contextServer.js (hunk 16 above)
Comment (jason-fox) [reply lib/services/northBound/contextServer.js]: Fixed  a5c650a

File: lib/services/northBound/contextServer.js (hunk 17 above)
Comment (jason-fox) [reply lib/services/northBound/contextServer.js]: Fixed: a5c650a

**Pull request merged**